tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,558 @@
+// cover all the added variables.
+func (gr *Graph) addInitializationGraphOp() {
+	inputs := make([]string, len(gr.variables))
+	i := 0
+	for input := range gr.variables {
+		inputs[i] = "^" + input + "/Assign"
+		i++
+	}
+
+	gr.def.Node = append(gr.def.Node, &pb.NodeDef{
+		Name:  "init",
+		Op:    "NoOp",
+		Input: inputs,
+	})
+}
+
+// Placeholder adds a placeholder to the Graph, a placeholder is an
+// operation that must be fed with data on execution.
+func (gr *Graph) Placeholder(name string, dataType DataType, dims []int64) *GraphNode {
+	op := &GraphNode{
+		outDataTypes: map[string]DataType{
+			name: dataType,
+		},
+		def: &pb.NodeDef{
+			Name: name,
+			Op:   "Placeholder",
+			Attr: make(map[string]*pb.AttrValue),
+		},
+	}
+	op.def.Attr["dtype"] = &pb.AttrValue{
+		Value: &pb.AttrValue_Type{
+			Type: pb.DataType(dataType),
+		},
+	}
+
+	shape := &pb.TensorShapeProto{
+		Dim: make([]*pb.TensorShapeProto_Dim, len(dims)),
+	}
+
+	for i, dim := range dims {
+		shape.Dim[i] = &pb.TensorShapeProto_Dim{
+			Size: dim,
+		}
+	}
+
+	op.def.Attr["shape"] = &pb.AttrValue{
+		Value: &pb.AttrValue_Shape{
+			Shape: shape,
+		},
+	}
+
+	gr.def.Node = append(gr.def.Node, op.def)
+
+	return op
+}
+
+// Marshal returns the current graph serialized so it can be exported.
+func (gr *Graph) Marshal() []byte {
+	result, _ := proto.Marshal(gr.def)
+
+	return result
+}
+
+// ErrExpectedVarAsinput is returned when the input value on an operation is
+// not a Variable and it must be a Variable.
+type ErrExpectedVarAsinput struct {
+	Operation string
+	InputPos  int
+}
+
+func (e *ErrExpectedVarAsinput) Error() string {
+	return fmt.Sprintf(
+		"The input value at pos %d for the operation '%s' must be of type Variable",
+		e.InputPos, e.Operation)
+}
+
+// ErrOperationNotFound is returned when the specified operation is not defined.
+type ErrOperationNotFound struct {
+	Op string
+}
+
+func (e *ErrOperationNotFound) Error() string {
+	return fmt.Sprintf("Operation '%s' not defined", e.Op)
+}
+
+// ErrInvalidAmounthOfInputs is returned when the number of inputs doesn't
+// corresponds with the expected for this operation.
+type ErrInvalidAmounthOfInputs struct {
+	Operation  string
+	OpInputs   int
+	SpecInputs int
+}
+
+func (e *ErrInvalidAmounthOfInputs) Error() string {
+	return fmt.Sprintf("Inputs required for operation '%s' %d, but %d provided",
+		e.Operation, e.OpInputs, e.SpecInputs)
+}
+
+// ErrMandatoryAttributeNotSpecified is returned when a mandatory attribute for
+// this operation was not specified.
+type ErrMandatoryAttributeNotSpecified struct {
+	Operation  string
+	AttribName string
+}
+
+func (e *ErrMandatoryAttributeNotSpecified) Error() string {
+	return fmt.Sprintf("The attribute '%s' is mandatory for operation '%s'",
+		e.AttribName, e.Operation)
+}
+
+// ErrInvalidAttrValue is returned when the data type of the value for this
+// attribute is not valid.
+type ErrInvalidAttrValue struct {
+	Operation  string
+	AttribName string
+}
```

Comment:
The term "Attribute" is inconsistently shortened in the errors -- it's called a MandatoryAttribute above, but here it's an Attr, and then inside the error it's an Attrib. :)  I suggest spelling it all the way out in all of the cases for now - ErrInvalidAttributeValue, AttributeName